replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/WorkerThreadPool.java

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.impl.worker;
+
+import com.connexta.replication.api.NodeAdapter;
+import com.connexta.replication.api.data.Site;
+import com.connexta.replication.api.impl.NodeAdapters;
+import com.connexta.replication.api.queue.Queue;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Manages a site's set of workers which in turn run workers.
+ */
+public class WorkerThreadPool {
+
+  private final Object lock = new Object();
+
+  private boolean shutdown = false;
+
+  private final List<Worker> workers = Collections.synchronizedList(new ArrayList<>());
+
+  private final NodeAdapters nodeAdapters;
+
+  private final Site localSite;
+
+  private final Site destinationSite;
+
+  private final Queue queue;
+
+  private AtomicInteger size = new AtomicInteger(0);
+
+  /**
+   * Creates a new worker thread pool for the given site and spins up a number of workers equal to
+   * the size passed in.
+   *
+   * @param queue queue which workers will poll tasks from
+   * @param size the initialize size of the pool
+   * @param localSite the local site
+   * @param remoteSite the remote site corresponding to the queue
+   * @param nodeAdapters factory for creating node adapters
+   */
+  public WorkerThreadPool(Queue queue, int size,
+      Site localSite,
+      Site remoteSite,
+      NodeAdapters nodeAdapters) {
+    this.queue = queue;
+    this.localSite = localSite;
+    this.destinationSite = remoteSite;
+    this.nodeAdapters = nodeAdapters;
+    setSize(size);
```

Comment:
❗️ A public method should never be called from within the constructor unless it is final

Reply:
Made it final for now, but i don't see this class ever being extended

Reply:
The point is that you should never call a method that can be overridden (wether you think it will or will never be) since you are calling into code of a class that hasn't been fully constructed yet. Even doing that somewhere in the constructor before all attributes have been initialized can create weird things since your method might be expecting an attribute to be null and it might be if it wasn't initialized yet. Even worst for derived classes since those have not be constructed yet and would get their overridden method called.